pennylane/tape/tape.py

```diff
@@ -195,16 +195,31 @@ def stop_at(obj):  # pylint: disable=unused-argument
                 new_queue.append(obj)
                 continue
 
-            if isinstance(obj, (Operator, qml.measurements.MeasurementProcess)):
-                # Object is an operation; query it for its expansion
-                try:
-                    obj = obj.expand()
-                except DecompositionUndefinedError:
-                    # Object does not define an expansion; treat this as
-                    # a stopping condition.
+            if isinstance(obj, Operator):
+                if not obj.has_decomposition:
+                    # Object does not define an expansion; treat this as a stopping condition.
                     new_queue.append(obj)
                     continue
 
+                decomp = make_qscript(obj.decomposition)()
```

Comment:
note that using queuing here can be important for certain things. for example, AmplitudeEmbedding queues to `_ops`, but it decomposes to QubitStateVector which queues to `_prep`

Reply:
second note: we used to catch `DecompositionUndefinedError`s here, but now we assume the decomposition will succeed if `has_decomposition` is True. This caused a test to fail with `QubitUnitary` because its `has_decomposition` was always returning True, despite it sometimes failing. I've updated it in this PR, but it is a change of behaviour.